design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,286 @@
+
+// InitGauge initializes a new gauge.
+// If a nil counter is returned during registeration, the method will
+// panic
+func InitGauge(r prometheus.Registerer, opts prometheus.GaugeOpts) prometheus.Gauge
+
+// InitHistogram initializes a new histogram.
+// If a nil counter is returned during registeration, the method will
+// panic
+func InitHistogram(r prometheus.Registerer, opts prometheus.HistogramOpts) prometheus.Histogram
+
+// registerCounterVec registers the CounterVec with the provided Registerer. If the
+// CounterVec has already been registered, the existing metric will be returned.
+func registerCounterVec(r prometheus.Registerer, g *prometheus.CounterVec) (*prometheus.CounterVec, error) 
+
+// registerHistogramVec registers the Histogram with the provided Registerer. If the
+// HistogramVec has already been registered, the existing metric will be returned.
+func registerHistogramVec(r prometheus.Registerer, g *prometheus.HistogramVec) (*prometheus.HistogramVec, error) 
+
+// registerGaugeVec registers the GaugeVec with the provided Registerer. If the
+// GaugeVec has already been registered, the existing GaugeVec will be returned.
+func registerGaugeVec(r prometheus.Registerer, g *prometheus.GaugeVec) (*prometheus.GaugeVec, error) 
+
+// registerGauge registers the Gauge with the provided Registerer. If the
+// gauge has already been registered, the existing metric will be returned.
+func registerGauge(r prometheus.Registerer, g *prometheus.Gauge) (*prometheus.Gauge, error) 
+
+// registerCounter registers the Counter with the provided Registerer. If the
+// counter has already been registered, the existing metric will be returned.
+func registerCounter(r prometheus.Registerer, g *prometheus.Counter) (*prometheus.Counter, error) 
+
+// registerHistogram registers the Histogram with the provided Registerer. If the
+// histogram has already been registered, the existing metric will be returned.
+func registerHistogram(r prometheus.Registerer, g *prometheus.Counter) (*prometheus.Histogram, error) 
+
+// registerCollector is an helper to register a metric and log registration errors
+func registerCollector(r prometheus.Registerer, c prometheus.Collector) (prometheus.Collector, error)
+```
+
+##### Example Initialization Steps for a new CounterVec metric
+1. Initialize a new CounterVec with relevant options and label names 
+
+2. Attempt to register the new CounterVec
+
+   a. If successful,
+
+       i. Generate combinations of label names
+
+       ii. Create counters for each combination and set the counter 
+           to 0.
+
+   b. If not successful, check if the error is an AlreadyRegisteredError
+
+       i. If yes, return the CounterVec and ignore the error
+
+       ii. If no, return a nil CounterVec and the received error. 
+
+3. If received a nil CounterVec from regitration, interrupt with a panic, because an interrupt would suggest a failure in the created CounterVec, which should be fixed by the programmer. 
```

Comment:
```suggestion
3. If received a nil CounterVec from registration, interrupt with a panic, because an interrupt would suggest a failure in the created CounterVec, which should be fixed by the programmer. 
```

Reply:
Fixed.